File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool validate(TreeNode* rt, long l, long r)
    {
        if(rt == nullptr)
            return true;
        if(!(rt -> val > l && rt -> val < r))
            return false;
        return validate(rt->left, l, rt -> val) && validate(rt -> right, rt -> val, r);
    }
    bool isValidBST(TreeNode* root) {
        return validate(root, -LONG_MAX, LONG_MAX);
    }
};

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def validate(root, leftVal, rightVal):
            if not root:
                return True
            if not (leftVal < root.val and rightVal > root.val):
                return False
            return validate(root.left, leftVal, root.val) and validate(root.right, root.val, rightVal)
        return validate(root, float('-inf'), float('inf'))